scarlet/component.py

```diff
@@ -128,86 +56,285 @@ def coord(self):
+    @property
+    def shift(self):
+        """Numpy view of the component center
+        """
+        if len(self._parameters) == 3:
+            return self._parameters[2]._data
+        return None
+
+    def get_model(self, *parameters):
         """Get the model for this component.
 
         Parameters
         ----------
-        sed: array
-            An sed to use in the model. If `sed` is `None`
-            then `self.sed` is used.
-        morph: array
-            A morphology to use in the model.
-            If `morph` is `None` then `self.morph`
-            is used.
+        parameters: tuple of optimimzation parameters
 
         Returns
         -------
-        model: array or tensor
-            (Bands, Height, Width) image of the model
-        """
-        if sed is not None and morph is not None:
-            return sed[:, None, None] * morph[None, :, :]
-        elif sed is None and morph is None:
-            return self._sed[:, None, None] * self._morph[None, :, :]
-        raise ValueError("You need to supply `sed` and `morph` or neither")
-
-    def get_flux(self):
-        """Get flux in every band
+        model: array
+            (Channels, Height, Width) image of the model
         """
-        return self.morph.sum() * self.sed
+        sed, morph, shift = None, None, None
+
+        # if params are set they are not Parameters, but autograd ArrayBoxes
+        # need to access the wrapped class with _value
+        for p in parameters:
+            if p._value is self._parameters[0]:
+                sed = p
+            if p._value is self._parameters[1]:
+                morph = p
+            if len(self._parameters) == 3 and p._value is self._parameters[2]:
+                shift = p
+
+        if sed is None:
+            sed = self.sed
+
+        if shift is None:
+            shift = self.shift
+
+        if morph is None:
+            # dont' use self._morph because we could have shift as parameter
+            morph =  self._pad_morph(self._shift_morph(shift, self._parameters[1]._data))
+        else:
+            morph =  self._pad_morph(self._shift_morph(shift, morph))
 
-    def backward_prior(self):
-        """Use the prior to update the gradient
-        """
-        if self.prior is not None:
-            self.prior.compute_grad(self)
-            if not self.fix_morph:
-                self.morph_grad += self.prior.morph_grad
-                self.L_morph += self.prior.L_morph
-            if not self.fix_sed:
-                self.sed_grad += self.prior.sed_grad
-                self.L_sed += self.prior.L_sed
-
-    def update(self):
-        """Update the component
-
-        This method can be overwritten in inherited classes to
-        run proximal operators or other component update functions
-        that will be executed during fitting.
-        """
-        return self
+        return sed[:, None, None] * morph[None, :, :]
+
+    def _pad_sed(self, sed):
```

Comment:
Fred is correct, since we do everything in full 3D, 1D slicing should be available. The operation of no-padding is super-cheap, so I haven't bothered. In fact, checking it led to an ambiguity (details...)